src/lib/Cashlink.ts

```diff
@@ -381,87 +379,35 @@ class Cashlink {
-                setTimeout(() => {
-                    this._executeUntilSuccess(fn, args).then(resolve);
-                }, 1000); // Retry in 1 second
-            });
-        }
-    }
-
     private async _getBlockchainHeight(): Promise<number> {
         await this._awaitConsensus();
-        return (await this._network).headInfo.height;
-    }
-
-    private async _getBalance(): Promise<number> {
-        if (this._balanceRequest) return this._balanceRequest;
-
-        const address = this.address.toUserFriendlyAddress();
-        const headHeight = await this._getBlockchainHeight();
-        return (this._balanceRequest = this._executeUntilSuccess<number>(async () => {
-            await this._awaitConsensus();
-            const balances = await (await this._network).getBalance(address);
-
-            // If the head changed in the meantime, it means the balance request got nulled. But code might still
-            // await this outdated promise, so we make sure to return the new promise from this old request.
-            if ((await this._network).headInfo.height !== headHeight && this._balanceRequest) {
-                return this._balanceRequest;
-            }
-
-            // Otherwise update the balance and resolve.
-            const balance = balances.get(address) || 0;
-            this._currentBalance = balance;
-            return balance;
-        }));
+        return (await this._getNetwork()).headInfo.height;
     }
 
     private async _onTransactionAddedOrRelayed(transaction: DetailedPlainTransaction): Promise<void> {
         if (transaction.recipient === this.address.toUserFriendlyAddress()
             || transaction.sender === this.address.toUserFriendlyAddress()) {
-            const amount = await this.getAmount(true);
-            this.fire(Cashlink.Events.UNCONFIRMED_BALANCE_CHANGE, amount);
             this.detectState();
         }
     }
 
-    private async _onHeadChanged(o: {height: number}): Promise<void> {
-        // balances potentially changed
-        this._balanceRequest = null;
-        // only interested in final balance
-        await this._onPotentialBalanceChange();
-    }
-
-    private async _onPotentialBalanceChange(): Promise<void> {
-        if ((await this._network).consensusState !== 'established') {
-            // only interested in final balance
-            return;
-        }
-        const oldBalance = this._currentBalance;
-        const balance = await this.getAmount();
-
-        if (balance !== oldBalance) {
-            this.fire(Cashlink.Events.CONFIRMED_BALANCE_CHANGE, balance);
+    private _onBalancesChanged(balances: Map<string, number>) {
+        const address = this.address.toUserFriendlyAddress();
+        if (balances.has(address)) {
+            this.balance = Nimiq.Policy.coinsToLunas(balances.get(address)!);
+            this.fire(Cashlink.Events.BALANCE_CHANGE, this.balance);
         }
-
-        this.detectState();
```

Comment:
I think if a transaction is mined quickly, it might happen that `TRANSACTION_PENDING` and `TRANSACTION_MINED` events do not reach us. Therefore, it would be good to keep the `detectState` check on balance updates (with the `if (balances.has(address))` check).
Might remove the event listener for `TRANSACTION_MINED` in turn.